src/utils/src/Threadpool.c

```diff
@@ -64,49 +70,63 @@ PVOID threadpoolActor(PVOID data)
+                        // otherwise exit loop and remove it.
+                        if (stackQueueGetCount(pThreadpool->threadList, &count) == STATUS_SUCCESS) {
+                            if (count > pThreadpool->minThreads) {
+                                finished = TRUE;
+                                if (stackQueueRemoveItem(pThreadpool->threadList, pThreadData) != STATUS_SUCCESS) {
+                                    DLOGE("Failed to remove thread data from threadpool");
+                                }
                             }
                         }
                     }
                 }
-            }
-            // this is a redundant safety check. To get here the threadpool must be being
-            // actively being destroyed, but it was unable to acquire our lock so entered a
-            // sleep spin check to allow this thread to finish. However somehow the task queue
-            // was not empty, so we ended up here. This check forces us to still exit gracefully
-            // in the event somehow the queue is not empty.
-            else if (ATOMIC_LOAD_BOOL(&pThreadData->terminate)) {
-                finished = TRUE;
-                if (stackQueueRemoveItem(pThreadpool->threadList, pThreadData) != STATUS_SUCCESS) {
-                    DLOGE("Failed to remove thread data from threadpool");
+                // this is a redundant safety check. To get here the threadpool must be being
+                // actively being destroyed, but it was unable to acquire our lock so entered a
+                // sleep spin check to allow this thread to finish. However somehow the task queue
+                // was not empty, so we ended up here. This check forces us to still exit gracefully
+                // in the event somehow the queue is not empty.
+                else if (ATOMIC_LOAD_BOOL(&pThreadData->terminate)) {
+                    finished = TRUE;
+                    if (stackQueueRemoveItem(pThreadpool->threadList, pThreadData) != STATUS_SUCCESS) {
+                        DLOGE("Failed to remove thread data from threadpool");
+                    }
                 }
+                MUTEX_UNLOCK(pThreadpool->listMutex);
+                MUTEX_UNLOCK(pThreadData->dataMutex);
             }
-            MUTEX_UNLOCK(pThreadpool->listMutex);
-            MUTEX_UNLOCK(pThreadData->dataMutex);
         } else {
             // couldn't lock our mutex, which means Threadpool locked this mutex to indicate
             // Threadpool has been deleted.
-            //
-            // Unlock mutex and free ThreadData
-            MUTEX_UNLOCK(pThreadData->dataMutex);
             finished = TRUE;
         }
     }
+    // now that we've released the listMutex, we can do an actual MUTEX_LOCK to ensure the
+    // threadpool has finished using pThreadData
+    MUTEX_LOCK(pThreadData->dataMutex);
+    MUTEX_UNLOCK(pThreadData->dataMutex);
```

Comment:
We are not doing anything between the lock and unlock?

Reply:
The comment explains it. We just want to grab the lock as a check that the threadpool constructor has finished using it. It's possible for us to fail the trylock in the loop, and then exit before the destructor has finished using pThreadData. If that happens, we could destroy it while they're trying to access it and segfault.